feat(build): add GitHub Action to make release

PsypherPunk opened the pull request:
Add config. to, when a new tag is pushed:

- build a `.deb` file using [`cargo-deb`](https://github.com/kornelski/cargo-deb);
- include the resulting `.deb` file in the subsequent release.

An example of one such release can be seen [here](https://github.com/PsypherPunk/RustScan/releases/tag/v6.6.6).

Relates #83

bee-san commented:
Merge conflicts :(

PsypherPunk commented:
@bee-san, I can look to resolve these but is there still interest in this PR?

bee-san commented:
> @bee-san, I can look to resolve these but is there still interest in this PR?

Oooh great question!

We have this release.yaml https://github.com/RustScan/RustScan/blob/master/.github/workflows/release.yml

But it does not do .deb , and we'd like to make it do .deb 😅 

I am thinking you can just copy this PR's yaml into that one?

PsypherPunk commented:
@bee-san, sure, I'll take a look. There's some nice matrix-build steps going on in there so I'll try to do something with it.

bee-san commented:
> @bee-san, sure, I'll take a look. There's some nice matrix-build steps going on in there so I'll try to do something with it.

Thank you, that would honestly be awesome. Before this we had a bulky Docker-builder to set up debs (feel free to delete https://github.com/RustScan/RustScan/tree/master/rustscan-debbuilder in your PR) 🙈 

Because of how cumbersome it was, new releases don't have .deb files :( 

I was just saying today one of the most important things I should add is .deb building automation!!

PsypherPunk commented on .github/workflows/release.yml:
```diff
@@ -121,6 +128,11 @@ jobs:
+
+          if [ "${{ matrix.build }}" = "x86_64-linux" ]; then
+            find target/ -name "*.deb"
+            cp "$(find target/ -name "*.deb")" "dist/"
```
This is _odd_.

You can see in [this](https://github.com/PsypherPunk/RustScan/actions/runs/8598981577/job/23560927701) run that the `find` here results in `target/x86_64-unknown-linux-gnu/debian/rustscan_2.1.1-1_amd64.deb`.

However, adding `cp target/x86_64-unknown-linux-gnu/debian/rustscan_*-1_amd64.deb` results in a `cp: cannot stat` error despite the `find` showing that it does exist…?!

PsypherPunk commented:
@bee-san, PR updated.

I've moved it to draft in case you have feedback; there's on odd line to which I've added a comment; it works—you can see the resulting release [here](https://github.com/PsypherPunk/RustScan/releases/tag/v6.6.6)—but there's probably a much better way to handle it (I could pull out the version as is done later and resolve the path to the `.deb` complete, I suppose).

PsypherPunk commented on .github/workflows/release.yml:
```diff
@@ -112,6 +112,13 @@ jobs:
+        if: matrix.build == 'x86_64-linux'
+        run: |
+          cargo install cargo-deb
+          cargo deb --target ${{ matrix.target }}
```
The `--deb-version` might be wiser here: elsewhere, versions are derived from the _tag_, not the Cargo version.

PsypherPunk replied:
Using `--deb-version` (which makes the path a little more deterministic too).

bee-san commented:
want me to review? looks ok to you? :)

PsypherPunk commented:
@bee-san, it works insofar as it generate a `.deb` file and attaches it to a release as per the existing `release.yml` so…yes 😬

As long as you're happy with the changes (and I deleted the right files you mentioned) then LGTM.

bee-san commented:
> @bee-san, it works insofar as it generate a `.deb` file and attaches it to a release as per the existing `release.yml` so…yes 😬
> 
> As long as you're happy with the changes (and I deleted the right files you mentioned) then LGTM.

Thanks!!

We only build for x86_64-linux? Do you think we should use the matrix to add support for more architectures or? 🤔

PsypherPunk commented:
I'm assuming it would only make sense to build `.deb` files for Linux and the only arch. is currently `x86_64-linux`…?

If that changes (say, adding `i686`), it should just be a matter of changing those `== 'x86_64-linux'` checks to `=~ 'linux'` or similar.

However, I'd probably consider doing that in a separate PR, once this one's proven with the existing arch./release-process…?

bee-san commented:
> I'm assuming it would only make sense to build `.deb` files for Linux and the only arch. is currently `x86_64-linux`…?
> 
> If that changes (say, adding `i686`), it should just be a matter of changing those `== 'x86_64-linux'` checks to `=~ 'linux'` or similar.
> 
> However, I'd probably consider doing that in a separate PR, once this one's proven with the existing arch./release-process…?

Ok thank you! GitHub's matrix sucks :P

bee-san commented:
It appears the release did not run? :( 

https://github.com/PsypherPunk/RustScan/actions/workflows/release.yml

```
name: Release
on:
  push:
    tags:
      - '[0-9]+.[0-9]+.[0-9]+'
  pull_request:
    branches:
      - master
    paths:
      - 'Cargo.toml'
      - 'Cargo.lock'
      - '.github/workflows/release.yml'
```

The latest tag is `2.2.2`, I am nto sure why it would not run :( 

https://github.com/RustScan/RustScan/releases/tag/2.2.2

PsypherPunk commented:
@bee-san, can I check how you actually created the tag? 

Not entirely sure if/why that would matter but I'll take a look 😬

bee-san commented:
> @bee-san, can I check how you actually created the tag?
> 
> Not entirely sure if/why that would matter but I'll take a look 😬

I made it in GitHub releases with 
![image](https://github.com/RustScan/RustScan/assets/10378052/8e1e474a-6cca-46b6-be3a-bb0fd9a2c8fe)

thank you for your contributions!

bee-san commented:
![image](https://github.com/RustScan/RustScan/assets/10378052/a80859b0-fa81-48e8-a6e7-8b399442674a)

Ok so it did run eventually, that's nice to know!! Docker and homebrew did not like it, stuff to fix 😂

PsypherPunk commented:
Phew! And there's a `.deb` an' everything!
